colossalai/zero/sharded_optim/sharded_optim_v2.py

```diff
@@ -87,24 +95,27 @@ def step(self, *args, **kwargs):
         # Write master param to payload
         for group in self.optim.param_groups:
             for p in group['params']:
-                if hasattr(p, 'ca_attr'):
-                    p.ca_attr.set_payload(p.data)
-                    p.data = p.ca_attr.payload()
+                is_param_sharded = p.col_attr.data.is_sharded
+                if not is_param_sharded:
+                    # The `p.col_attr.data` saves full fp16 param
+                    # But we only have updated fp32 param shard here
+                    # So we first shard full fp16 param and copy fp32 param shard to it
+                    # Then we will gather them
+                    self.shard_strategy.shard([p.col_attr.data])
+                p.col_attr.data.copy_payload(p.data)
```

Comment:
Just copy the valid piece of p.data to p.col_attr and no shard and gather required. It can reduce some time in __init___ but not significant to the overall training performance as the init just executed once.
You can add a TODO for such optimization and we can do it latter.

Reply:
I will discuss you in the office later.